Source/Core/VideoBackends/D3D12/D3DQueuedCommandList.cpp

```diff
@@ -0,0 +1,1217 @@
+					// http://stackoverflow.com/questions/1420029/how-to-break-out-of-a-loop-from-inside-a-switch
+
+					item += sizeof(D3DQueueItemType) * 2;
+
+					if (item - queue_array > s_queue_array_size / 3)
+					{
+						item = queue_array;
+					}
+
+					goto exitLoop;
+			}
+		}
+
+		exitLoop:
+
+		parent_queued_command_list->m_queue_array_front = static_cast<UINT>(item - queue_array);
+	}
+}
+
+ID3D12QueuedCommandList::ID3D12QueuedCommandList(ID3D12GraphicsCommandList* backing_command_list, ID3D12CommandQueue* backing_command_queue) :
+	m_command_list(backing_command_list),
+	m_command_queue(backing_command_queue)
+{
+	memset(m_queue_array, 0, sizeof(m_queue_array));
+
+	m_queue_array_back = m_queue_array;
+
+	m_begin_execution_event = CreateSemaphore(nullptr, 0, 256, nullptr);
+
+	m_background_thread = CreateThread(nullptr, 0, BackgroundThreadFunction, this, 0, &m_background_thread_id);
+}
+
+ID3D12QueuedCommandList::~ID3D12QueuedCommandList()
+{
+	TerminateThread(m_background_thread, 0);
+	CloseHandle(m_background_thread);
+
+	CloseHandle(m_begin_execution_event);
+}
+
+void ID3D12QueuedCommandList::QueueExecute()
+{
+	reinterpret_cast<D3DQueueItem*>(m_queue_array_back)->Type = D3DQueueItemType::ExecuteCommandList;
+
+	m_queue_array_back += sizeof(ExecuteCommandListArguments) + sizeof(D3DQueueItemType) * 2;
+}
+
+void ID3D12QueuedCommandList::QueueFenceGpuSignal(ID3D12Fence* fence_to_signal, UINT64 fence_value)
+{
+	D3DQueueItem item = {};
+
+	item.Type = D3DQueueItemType::FenceGpuSignal;
+	item.FenceGpuSignal.fence = fence_to_signal;
+	item.FenceGpuSignal.fence_value = fence_value;
+
+	*reinterpret_cast<D3DQueueItem*>(m_queue_array_back) = item;
+
+	m_queue_array_back += sizeof(FenceGpuSignalArguments) + sizeof(D3DQueueItemType) * 2;
+}
+
+void ID3D12QueuedCommandList::QueueFenceCpuSignal(ID3D12Fence* fence_to_signal, UINT64 fence_value)
+{
+	D3DQueueItem item = {};
+
+	item.Type = D3DQueueItemType::FenceCpuSignal;
+	item.FenceCpuSignal.fence = fence_to_signal;
+	item.FenceCpuSignal.fence_value = fence_value;
+
+	*reinterpret_cast<D3DQueueItem*>(m_queue_array_back) = item;
+
+	m_queue_array_back += sizeof(FenceCpuSignalArguments) + sizeof(D3DQueueItemType) * 2;
+}
+
+void ID3D12QueuedCommandList::QueuePresent(IDXGISwapChain* swap_chain, UINT sync_interval, UINT flags)
+{
+	D3DQueueItem item = {};
+
+	item.Type = D3DQueueItemType::Present;
+	item.Present.swapChain = swap_chain;
+	item.Present.flags = flags;
+	item.Present.syncInterval = sync_interval;
+
+	*reinterpret_cast<D3DQueueItem*>(m_queue_array_back) = item;
+
+	m_queue_array_back += sizeof(PresentArguments) + sizeof(D3DQueueItemType) * 2;
+}
+
+void ID3D12QueuedCommandList::ClearQueue()
+{
+	// Drain semaphore to ensure no new previously queued work executes (though inflight work may continue).
+	while (WaitForSingleObject(m_begin_execution_event, 0) != WAIT_TIMEOUT) { }
+
+	// Assume that any inflight queued work will complete within 100ms. This is a safe assumption.
+	Sleep(100);
```

Comment:
Not sure if this is a good idea? (again, stalls?) Perhaps use an event, to indicate the status of the worker?